### PR TITLE
State saving, see #283

### PR DESCRIPTION
**Pull request by SwxtchCode**
## Description
Addressing #283 and #318. Adds ability to store player data locally so the user can restore previous player history whenever they want. Currently unfinished, making this a draft.

## Screenshots
Will add at a later date.

### CODE REVIEW

**Comment by SwxtchCode** [comment]
it should work but i <s>dont want to</s> dont have time to test

**Comment by BoredFishRE** [comment]
whoops wrong button not ready for review ignore that

**Comment by BoredFishRE** [comment]
there we go

**Comment by SwxtchCode** [comment]
i think  i should test this but webpack doesnt work for me

**Comment by ArcerionDev** [comment]
https://github.com/Prodigy-Hacking/ProdigyMathGameHacking/pull/349/commits/69a7480e83e0ad2d1021c15b4008d1a2bbfc54e5#diff-e220704dd9f1e93103e38cf2f64ffe6e38ab6fb7340616d3646b90a1423d7772L40
Why did you just add a space

**Comment by BoredFishRE** [comment]
Because yes

**Comment by ArcerionDev** [comment]
I've found it's not possible to store objects in localStorage, it has to be a primitive value.

**Comment by PatheticMustan** [comment]
> therefore, this may be a bit difficult.

I don't see what the issue is, just JSON.stringify and JSON.parse it

**Comment by PatheticMustan** [comment]
@SwxtchCode, have you tested it yet?

File: willsCheatMenu/src/hacks/misc.ts [hunk 1]
```diff
@@ -2,70 +2,132 @@ import { Swal, Toast, NumberInput, Confirm } from "../utils/swal";
...
+new Hack(category.misc, "Reset Account", "Completely resets your account.").setClick(async () => {
+  if (!(await Confirm.fire("Are you sure you want to reset your account?", "This action is not reversable.")).value) return;
+  _.player.resetAccount();
+});
+
+new Hack(category.misc, "Restore Player State", "Restores your player data to an earlier state (WCM must've been installed at the time, and the history data is local to this device only.)").setClick(async () => {
+  playerStateHistory = JSON.parse(localStorage.getItem("playerStateHistory")) || [];
+  
+  inputOptionsPromise = new Promise(function(resolve) {
+    setTimeout(function() {
```
Comment (PatheticMustan) [inline willsCheatMenu/src/hacks/misc.ts:104]: Can I have more info about why you used setTimeout?

**Comment by SwxtchCode** [comment]
webpack doesnt work for me

**Comment by PatheticMustan** [comment]
I'll tell you how to locally set up redirector and help u fix webpack later. For now, it looks good.

Just add a save state button and it should be fine.

**Comment by ArcerionDev** [comment]
> > therefore, this may be a bit difficult.
> 
> I don't see what the issue is, just JSON.stringify and JSON.parse it

you can't stringify circular objects, and _.player is.

File: willsCheatMenu/src/hacks/misc.ts (hunk 1 above)
Comment (SwxtchCode) [reply willsCheatMenu/src/hacks/misc.ts:104]: i copied something from stackoverflow and it had a `setTimeout` thing

**Comment by SwxtchCode** [comment]
> > > therefore, this may be a bit difficult.
> > 
> > I don't see what the issue is, just JSON.stringify and JSON.parse it

> you can't stringify circular objects, and _.player is.

shoot i see

**Comment by SwxtchCode** [comment]
> 69a7480
> Why did you just add a space

because i was trying to change it to two-space indent but failed. i prefer 2 space indent.

**Comment by SwxtchCode** [comment]
i forgot this exist

**Pull request closed without merging**